# Trajectory Data Visualization

This notebook visualizes the preprocessed trajectory data:
- **train_x.pt**: 2-second history (past positions)
- **train_y.pt**: 6-second future (what we want to predict)

All coordinates are **ego-centric** (centered on the car, rotated to face north).

## Step 1: Import Libraries

In [3]:
import torch
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.patches import Circle, FancyArrowPatch
import warnings
warnings.filterwarnings('ignore')

print("✓ Libraries imported")

✓ Libraries imported


## Step 2: Load the Tensor Files

In [ ]:
# Load training data
train_x = torch.load('/media/robotswithai/data1/ece1508/train_x.pt')
train_y = torch.load('/media/robotswithai/data1/ece1508/train_y.pt')

print(f"✓ Data loaded!")
print(f"\ntrain_x shape: {train_x.shape}")
print(f"  - {train_x.shape[0]} trajectories")
print(f"  - {train_x.shape[1]} frames (2 sec at 2Hz)")
print(f"  - {train_x.shape[2]} coordinates (x, y)")

print(f"\ntrain_y shape: {train_y.shape}")
print(f"  - {train_y.shape[0]} trajectories")
print(f"  - {train_y.shape[1]} frames (6 sec at 2Hz)")
print(f"  - {train_y.shape[2]} coordinates (x, y)")

✓ Data loaded!

train_x shape: torch.Size([231909, 4, 2])
  - 231909 trajectories
  - 4 frames (2 sec at 2Hz)
  - 2 coordinates (x, y)

train_y shape: torch.Size([231909, 12, 2])
  - 231909 trajectories
  - 12 frames (6 sec at 2Hz)
  - 2 coordinates (x, y)


: 

## Step 3: Visualize One Trajectory

In [ ]:
# Pick the first trajectory
idx = 0
history = train_x[idx].numpy()  # Shape: (4, 2)
future = train_y[idx].numpy()   # Shape: (12, 2)

print(f"Trajectory {idx}:")
print(f"\nHistory (past 2 seconds):")
for i, pos in enumerate(history):
    print(f"  Frame {i}: x={pos[0]:7.2f}, y={pos[1]:7.2f}")

print(f"\nFuture (next 6 seconds):")
for i, pos in enumerate(future):
    print(f"  Frame {i}: x={pos[0]:7.2f}, y={pos[1]:7.2f}")

## Step 4: Plot the Trajectory

In [ ]:
fig, ax = plt.subplots(figsize=(10, 8))

# Plot history (blue)
ax.plot(history[:, 0], history[:, 1], 'b-o', linewidth=2, markersize=8, label='History (past 2 sec)')
ax.scatter(history[0, 0], history[0, 1], color='blue', s=200, marker='s', label='Start', zorder=5)

# Plot future (red)
ax.plot(future[:, 0], future[:, 1], 'r-o', linewidth=2, markersize=8, label='Future (next 6 sec)')
ax.scatter(future[-1, 0], future[-1, 1], color='red', s=200, marker='*', label='End', zorder=5)

# Draw ego vehicle at origin (0, 0) facing north (up)
circle = Circle((0, 0), 0.5, color='green', fill=False, linewidth=2, label='Ego vehicle')
ax.add_patch(circle)
ax.arrow(0, 0, 0, 1.5, head_width=0.3, head_length=0.3, fc='green', ec='green')

# Styling
ax.set_xlabel('X Position (meters)', fontsize=12)
ax.set_ylabel('Y Position (meters)', fontsize=12)
ax.set_title(f'Trajectory {idx}: Vehicle Motion\n(Ego-centric frame)', fontsize=14, fontweight='bold')
ax.grid(True, alpha=0.3)
ax.legend(fontsize=11, loc='best')
ax.set_aspect('equal')

plt.tight_layout()
plt.show()

print("✓ Plot complete!")

## Step 5: Show Multiple Examples

In [ ]:
# Plot 6 random trajectories
fig, axes = plt.subplots(2, 3, figsize=(15, 10))
axes = axes.flatten()

np.random.seed(42)
indices = np.random.choice(len(train_x), 6, replace=False)

for subplot_idx, traj_idx in enumerate(indices):
    ax = axes[subplot_idx]
    
    history = train_x[traj_idx].numpy()
    future = train_y[traj_idx].numpy()
    
    # Plot
    ax.plot(history[:, 0], history[:, 1], 'b-o', linewidth=2, markersize=6, label='History')
    ax.plot(future[:, 0], future[:, 1], 'r-o', linewidth=2, markersize=6, label='Future')
    
    # Ego vehicle
    circle = Circle((0, 0), 0.3, color='green', fill=False, linewidth=1.5)
    ax.add_patch(circle)
    ax.arrow(0, 0, 0, 0.8, head_width=0.2, head_length=0.2, fc='green', ec='green', alpha=0.7)
    
    ax.set_xlabel('X (m)', fontsize=10)
    ax.set_ylabel('Y (m)', fontsize=10)
    ax.set_title(f'Trajectory {traj_idx}', fontsize=11, fontweight='bold')
    ax.grid(True, alpha=0.3)
    ax.set_aspect('equal')
    ax.legend(fontsize=9)

plt.suptitle('Sample Trajectories from Dataset', fontsize=14, fontweight='bold')
plt.tight_layout()
plt.show()

print("✓ 6 example trajectories plotted")

## Step 6: Data Statistics

In [ ]:
print("=" * 60)
print("DATA STATISTICS")
print("=" * 60)

print(f"\nTotal trajectories: {train_x.shape[0]:,}")
print(f"Data type: {train_x.dtype}")
print(f"Memory usage: {train_x.element_size() * train_x.nelement() / 1e6:.1f} MB (train_x)")
print(f"             {train_y.element_size() * train_y.nelement() / 1e6:.1f} MB (train_y)")

print(f"\n--- HISTORY (train_x) ---")
print(f"Shape: {train_x.shape}")
print(f"X range: [{train_x[:, :, 0].min():.2f}, {train_x[:, :, 0].max():.2f}] meters")
print(f"Y range: [{train_x[:, :, 1].min():.2f}, {train_x[:, :, 1].max():.2f}] meters")
print(f"Mean position: ({train_x[:, :, 0].mean():.2f}, {train_x[:, :, 1].mean():.2f})")
print(f"Std deviation: ({train_x[:, :, 0].std():.2f}, {train_x[:, :, 1].std():.2f})")

print(f"\n--- FUTURE (train_y) ---")
print(f"Shape: {train_y.shape}")
print(f"X range: [{train_y[:, :, 0].min():.2f}, {train_y[:, :, 0].max():.2f}] meters")
print(f"Y range: [{train_y[:, :, 1].min():.2f}, {train_y[:, :, 1].max():.2f}] meters")
print(f"Mean position: ({train_y[:, :, 0].mean():.2f}, {train_y[:, :, 1].mean():.2f})")
print(f"Std deviation: ({train_y[:, :, 0].std():.2f}, {train_y[:, :, 1].std():.2f})")

print("\n" + "=" * 60)

## Step 7: Verify Data Quality

In [ ]:
print("\n✓ Checking data quality...\n")

# Check for NaNs
nan_x = torch.isnan(train_x).sum().item()
nan_y = torch.isnan(train_y).sum().item()
print(f"NaN values in train_x: {nan_x}")
print(f"NaN values in train_y: {nan_y}")

if nan_x == 0 and nan_y == 0:
    print("✓ No NaN values - data is clean!")

# Check trajectory continuity (movement distance)
distances_x = []
for i in range(min(1000, len(train_x))):
    history = train_x[i].numpy()
    # Distance from first to last frame
    dist = np.linalg.norm(history[-1] - history[0])
    distances_x.append(dist)

print(f"\nHistory movement (first to last frame):")
print(f"  Mean: {np.mean(distances_x):.2f} meters")
print(f"  Min: {np.min(distances_x):.2f} meters")
print(f"  Max: {np.max(distances_x):.2f} meters")

print("\n✓ Data validation complete!")

## Summary

You now have:
- **231,909 trajectories** ready for training
- Each trajectory has **4 frames of history** (2 seconds at 2Hz)
- Each trajectory has **12 frames of future** (6 seconds at 2Hz)
- All coordinates are **ego-centric** (car at origin facing north)

### Next Steps:
1. ✅ **Preprocess** (already done - `preprocess.py`)
2. ✅ **Visualize** (this notebook)
3. **Train Transformer** (next: `train.py`)
4. **Evaluate** (plot predictions vs ground truth)